crates/miden-tx/src/executor/mod.rs

```diff
@@ -360,31 +360,41 @@ where
         .map_err(TransactionExecutorError::TransactionProgramExecutionFailed);
 
         match result {
-            Ok(_) => Ok(NoteAccountExecution::Success),
-            Err(tx_execution_error) => {
+            Ok(_) => Ok(NoteConsumptionInfo::new_successful(
+                tx_inputs
+                    .into_input_notes()
+                    .into_iter()
+                    .map(|note| note.into_note())
+                    .collect::<Vec<_>>(),
+            )),
+            Err(error) => {
+                // Map the last note id from execution to the failed note.
                 let notes = host.tx_progress().note_execution();
-
-                // empty notes vector means that we didn't process the notes, so an error
-                // occurred somewhere else
-                if notes.is_empty() {
-                    return Err(tx_execution_error);
-                }
-
-                let ((last_note, last_note_interval), success_notes) = notes
+                let ((last_note, _last_note_interval), success_notes) = notes
                     .split_last()
                     .expect("notes vector should not be empty because we just checked");
-
-                // if the interval end of the last note is specified, then an error occurred after
-                // notes processing
-                if last_note_interval.end().is_some() {
-                    return Err(tx_execution_error);
-                }
-
-                Ok(NoteAccountExecution::Failure {
-                    failed_note_id: *last_note,
-                    successful_notes: success_notes.iter().map(|(note, _)| *note).collect(),
-                    error: Some(tx_execution_error),
-                })
+                let note = input_notes
+                    .iter()
+                    .find(|&note| note.id() == *last_note)
+                    .expect("Last note returned from note execution should exist in input notes")
+                    .note()
+                    .clone();
+                let failed = vec![NoteConsumptionError::ExecutionError { note, error }];
+                // Gather successful notes.
+                let success_notes = success_notes.iter().map(|(id, _)| *id).collect::<Vec<_>>();
+                let successful = tx_inputs
+                    .into_input_notes()
+                    .into_iter()
+                    .filter_map(|note| {
+                        // O(n*m) is fine for the input sizes we deal with here.
+                        if success_notes.contains(&note.id()) {
+                            Some(note.into_note())
+                        } else {
+                            None
+                        }
+                    })
+                    .collect::<Vec<_>>();
+                Ok(NoteConsumptionInfo::new(successful, failed))
```

Comment:
This is a bit difficult to follow. I would potentially move this code into a specialized constructor of `NoteConsumptionInfo`.

Reply:
I don't think it makes sense to put in a constructor because its mapping note Ids to notes which `NoteConsumptionInfo` is not concerned with.

I could pull it out into an associated fn but I think it would be an awkward set of arguments for it.